src/main/java/io/jenkins/jenkins/plugins/okhttp/api/JenkinsOkHttpClient.java

```diff
@@ -0,0 +1,172 @@
+                @Override
+                public List<Proxy> select(URI uri) {
+                    final List<Proxy> proxies = new ArrayList<>(1);
+                    String host = uri.getHost();
+                    for (Pattern p : proxy.getNoProxyHostPatterns()) {
+                        if (p.matcher(host).matches()) {
+                            proxies.add(Proxy.NO_PROXY);
+                            return proxies;
+                        }
+                    }
+                    proxies.add(proxyServer);
+                    return proxies;
+                }
+
+                @Override
+                public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+                    LOGGER.log(Level.WARNING, "Proxy connection failed", ioe);
+                }
+            };
+            reBuild.proxySelector(proxySelector);
+        } else {
+            reBuild.proxy(Proxy.NO_PROXY);
+        }
+
+        if (acceptAnyCertificate) {
+            try {
+                applyInsecureConfiguration(reBuild);
+            } catch (NoSuchAlgorithmException | KeyManagementException e) {
+                LOGGER.log(Level.WARNING, "Cannot apply insecure trust all certs configuration to okhttp", e);
+            }
+        }
+        return reBuild;
+    }
+
+    /**
+     * Updates non-proxy-hosts on the given client (returns a new builder object as a rebuild of the given one).
+     * @param httpClient the base client
+     * @param nonProxyHosts list of non proxy hosts patterns
+     * @return the builder for the new client
+     */
+    public static OkHttpClient.Builder withNonProxyHosts(OkHttpClient httpClient, List<String> nonProxyHosts) {
+        ProxySelector proxySelector = httpClient.proxySelector();
+        List<Pattern> noProxyHostPatterns = ProxyConfiguration.getNoProxyHostPatterns(String.join(",", nonProxyHosts));
+        OkHttpClient.Builder reBuild = httpClient.newBuilder();
+        reBuild.proxySelector(new ProxySelector() {
+            @Override
+            public List<Proxy> select(URI uri) {
+                final List<Proxy> proxies = new ArrayList<>(1);
+                String host = uri.getHost();
+                for (Pattern p : noProxyHostPatterns) {
+                    if (p.matcher(host).matches()) {
+                        proxies.add(Proxy.NO_PROXY);
+                        return proxies;
+                    }
+                }
+                return proxySelector.select(uri);
+            }
+
+            @Override
+            public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+                LOGGER.log(Level.WARNING, "Proxy connection failed", ioe);
+            }
+        });
+        return reBuild;
+    }
+
+    private static void applyInsecureConfiguration(OkHttpClient.Builder reBuild) throws NoSuchAlgorithmException, KeyManagementException {
+        final TrustManager[] trustAllCerts = new TrustManager[] {
```

Comment:
wondering why creating a array when L164 is only using the first one.

Reply:
Because on L163 we need an array.

Reply:
Didn't see that usage. Thanks.